### DISCUSSION TITLE
strange error: `/usr/include/linux/videodev2.h:59:10: fatal error: 'sys/time.h' file not found`

### DISCUSSION THREAD

**Discussion by makerio90** [cross-rs/cross#1129] [Q&A] [answered]
I'm trying to build the `v4l2-sys-mit` dependency of `v4l2`, and am  getting a strange build error. `v4l2-sys-mit` has a build script witch binds `/usr/include/linux/videodev2.h`, witch binds `sys/time.h`. whenever i try and compile i get this error
```
Caused by:
  process didn't exit successfully: `/target/debug/build/v4l2-sys-mit-93e44231da6649fa/build-script-build` (exit status: 101)
  --- stderr
  /usr/include/linux/videodev2.h:59:10: fatal error: 'sys/time.h' file not found
  /usr/include/linux/videodev2.h:59:10: fatal error: 'sys/time.h' file not found, err: true
  thread 'main' panicked at 'Failed to generate bindings: ()', /cargo/registry/src/github.com-1ecc6299db9ec823/v4l2-sys-mit-0.2.0/build.rs:10:10
```
i have no idea what I'm doing any help appreciated

**Comment by Emilgardis** [comment]
what target are you building for? What version of cross are you using?

**Reply by makerio90** [reply]
woops srry;

target: `aarch64-unknown-linux-gnu`
version: v0.2.4

**Reply by Emilgardis** [reply]
how are you installing the libraries needed? i.e what image are you using and what's your `Cross.toml` configuration?

**Reply by makerio90** [reply]
im using default other than a run hook in my Cargo.toml
```
[package.metadata.cross.target.aarch64-unknown-linux-gnu]
pre-build = ["dpkg --add-architecture arm64 ", "apt-get update ", "apt-get install --assume-yes OpenSSL"]
```

**Reply by Alexhuszagh** [reply]
`openssl` isn't the package you want: it's `libssl-dev`, and also, you need to install it for the correct architecture: `CROSS_DEB_ARCH`. You take a look at the [contents](https://packages.ubuntu.com/jammy/amd64/openssl/filelist) of the package for the architecture you're building on, it has none of the development files: only the OpenSSL binary:

```
/etc/ssl/openssl.cnf
/usr/bin/c_rehash
/usr/bin/openssl
/usr/lib/ssl/certs
/usr/lib/ssl/misc/CA.pl
/usr/lib/ssl/misc/tsget
/usr/lib/ssl/misc/tsget.pl
/usr/lib/ssl/openssl.cnf
/usr/lib/ssl/private
...
```

**Reply by Alexhuszagh** [reply]
This is the package you want: [libssl-dev:arm64](https://packages.ubuntu.com/focal/arm64/libssl-dev/filelist):

```
/usr/include/aarch64-linux-gnu/openssl/opensslconf.h
/usr/include/openssl/aes.h
/usr/include/openssl/asn1.h
/usr/include/openssl/asn1_mac.h
/usr/include/openssl/asn1err.h
/usr/include/openssl/asn1t.h
/usr/include/openssl/async.h
/usr/include/openssl/asyncerr.h
/usr/include/openssl/bio.h
/usr/include/openssl/bioerr.h
/usr/include/openssl/blowfish.h
/usr/include/openssl/bn.h
/usr/include/openssl/bnerr.h
/usr/include/openssl/buffer.h
/usr/include/openssl/buffererr.h
/usr/include/openssl/camellia.h
...
```

**Reply by Alexhuszagh** [reply]
Finally (sorry for the multiple posts, they're somewhat separate topics), the compiler should be detecting the correct include directory unless something is going very off in the build system:

```bash
$ find /usr/aarch64-linux-gnu/ -name time.h | grep sys
/usr/aarch64-linux-gnu/include/sys/time.h
$ echo $CROSS_SYSROOT/
/usr/aarch64-linux-gnu/
$ echo '#include <sys/time.h>
int main() { return 0; }' > main.c
$ $CC_aarch64_unknown_linux_gnu main.c
```

And everything compiles as expected. This suggests something in the build system isn't correctly supporting custom toolchains, and is using something hardcoded or overriding the include paths, or you're missing necessary dependencies.

I think you need to include [libv4l-dev:arm64](https://packages.ubuntu.com/jammy/libv4l-dev).

**Reply by makerio90** [reply]
openssl is for a diffrent crate and `$CROSS_SYSROOT` is not set.

**Reply by Alexhuszagh** [reply]
> openssl is for a diffrent crate and `$CROSS_SYSROOT` is not set.

You're using an older version of the image then. But even if it isn't, it's just a helper: the include search paths will still find the `<sys/time.h>` regardless from the cross-compiler. Also: I sincerely doubt you want the binary utility of OpenSSL no matter what crate you are building for: you want `libssl-dev` for the architecture you are building for, in this case, arm64.

**Reply by Emilgardis** [reply]
if a build script needs openssl you'd need to install it as you have done

but as @Alexhuszagh points out, you need to install the library needed for `v4l2-sys-mit`

```
[dependencies]
v4l2-sys-mit = "0.2.0"

[package.metadata.cross.target.aarch64-unknown-linux-gnu]
pre-build = [
    "dpkg --add-architecture arm64 ",
    "apt-get update ",
    "apt-get install --assume-yes libv4l-dev:arm64",
]
```

with just the above it seems to compile correctly

**Reply by makerio90** [reply]
im using `libv4l-dev:arm64` and `openssl-dev:arm64` (ive removed the bin package) and im still getting the same error, i may just keep trying to use the native build instead of cross.

thx anyway

**Reply by Emilgardis** [reply]
Would you mind posting all dependencies? 

Because this should work

using the latest cross with the 0.2.4 image you're using

```toml
[package]
name = "cross-1129"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
v4l2-sys-mit = "0.2.0"

[package.metadata.cross.target.aarch64-unknown-linux-gnu]
image = "ghcr.io/cross-rs/aarch64-unknown-linux-gnu:0.2.4"
pre-build = [
    "dpkg --add-architecture arm64 ",
    "apt-get update ",
    "apt-get install --assume-yes libv4l-dev:arm64",
]
```

```
❯ cross build --target aarch64-unknown-linux-gnu
[+] Building 0.1s (6/6)
 => [internal] load build definition from Dockerfile.aarch64-unknown-linux-gnu-custom                                                                                 0.0s
[+] Building 0.1s (6/6) FINISHED
 => [internal] load build definition from Dockerfile.aarch64-unknown-linux-gnu-custom                                                                                 0.0s
 => => transferring dockerfile: 247B                                                                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                     0.0s
 => => transferring context: 2B                                                                                                                                       0.0s
 => [internal] load metadata for ghcr.io/cross-rs/aarch64-unknown-linux-gnu:main                                                                                      0.0s
 => [1/2] FROM ghcr.io/cross-rs/aarch64-unknown-linux-gnu:main                                                                                                        0.0s
 => CACHED [2/2] RUN eval "dpkg --add-architecture arm64
apt-get update o image                                                                                                                                                0.0s
apt-get install --assume-yes libv4l-dev:arm64"                                                                                                                        0.0s
 => exporting to image                                                                                                                                                0.0s
 => => exporting layers                                                                                                                                               0.0s
 => => writing image sha256:362b74d1c10b44fe8d25f7753aabeb941cdf98e02b827feb33f1185e9427702e                                                                          0.0s
 => => naming to localhost/cross-rs/cross-custom-cross-1129:aarch64-unknown-linux-gnu-59120-pre-build                                                                 0.0s
   Compiling v4l2-sys-mit v0.2.0
   Compiling cross-1129 v0.1.0 (/Users/emil/workspace/cross-tests/cross-1129)
    Finished dev [unoptimized + debuginfo] target(s) in 3m 49s
```

**Comment by x-paranoid-android** [comment, marked as answer]
Just had the same problem. Here is solution for main image:
```
[package.metadata.cross.target.aarch64-unknown-linux-gnu]
image = "ghcr.io/cross-rs/aarch64-unknown-linux-gnu:main"
pre-build = [
    "dpkg --add-architecture arm64 ",
    "apt-get update ",
    "apt-get install --assume-yes libc6-dev:arm64",
]
```
and for 0.2.4 image:
```
[package.metadata.cross.target.aarch64-unknown-linux-gnu]
image = "ghcr.io/cross-rs/aarch64-unknown-linux-gnu:0.2.4"
pre-build = [
    "dpkg --add-architecture arm64 ",
    "apt-get update ",
    "apt-get install --assume-yes libc6-dev:arm64 libclang-3.9",
]
```

**Reply by Alexhuszagh** [reply]
I think this is the general answer, just substituting `libc6-dev` and `libclang-3.9` (the latter I'm guessing is for bindgen or similar) with whatever packages necessary for the target.

**Comment by dkapnopoulos** [comment]
I have the same problem with the same library and almost the same setup. I don't use cross but doing the cross compile manually with Docker.
The proposed solution did not solve the problem for me.
Since I have all the required libraries installed, I concluded that the problem is the include path of bindgen.
Thus, I added this line in my Docker file:
`ENV CPATH="${CPATH}:/usr/aarch64-linux-gnu/include"`

That solved my problem (although not very elegant :P)!

**Reply by apbr** [reply]
I had the same problem and found a bit more elegant solution by setting `BINDGEN_EXTRA_CLANG_ARGS='--sysroot /usr/aarch64-linux-gnu'`.
Maybe this works for you as well.
